kernel/queue.c

```diff
@@ -207,19 +207,23 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
 
 	event.state = K_POLL_STATE_NOT_READY;
 
-	err = k_poll(&event, 1, timeout);
-	if (err) {
-		return NULL;
-	}
+	do {
+		err = k_poll(&event, 1, timeout);
+		if (err) {
+			return NULL;
+		}
 
-	__ASSERT_NO_MSG(event.state == K_POLL_STATE_FIFO_DATA_AVAILABLE);
+		__ASSERT_NO_MSG(event.state ==
+				K_POLL_STATE_FIFO_DATA_AVAILABLE);
+
+		/* sys_slist_* aren't threadsafe, so must be always protected by
+		 * irq_lock.
+		 */
+		key = irq_lock();
+		val = sys_slist_get(&queue->data_q);
+		irq_unlock(key);
+	} while(!val && timeout == K_FOREVER);
```

Comment:
Prefer a onetime if-clause to check timeout == K_FOREVER, avoids extra branch-instructions in a loop.